### PR TITLE
proposal for custom field management

### PR DESCRIPTION
**Pull request by stivni**
TODO:
- [x] groups
- [ ] change type `customField` to `customFieldDefinition` everywhere

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -2793,6 +2793,180 @@ Register a new webhook.
...
+
+### customFieldDefinitions.list [GET /customFieldDefinitions.list]
+
+Get a list of all the definitions of custom fields.
+
++ Request (application/json;charset=utf-8)
+
+    + Attributes (object)
+        + page (Page, optional)
+        + context: `contact` (enum[string], required) - The context of the custom field definition
```
Comment (Brammm) [inline apiary.apib]: Q: I might have missed previous discussion, but are we settling on `context`? Is `for` still used somewhere else? Probably should be consistent across all of TL.

File: apiary.apib [hunk 2]
```diff
@@ -2793,6 +2793,180 @@ Register a new webhook.
...
+            + type (enum[string])
+                + Members
+                    + number
+                    + auto_increment
+                    + string
+                    + enum
+                    + user
+                    + company
+                    + contact
+                    + ___more___
```
Comment (Brammm) [inline apiary.apib]: Q: what's the reason for not providing the full list here?

File: apiary.apib [hunk 3]
```diff
@@ -2793,6 +2793,180 @@ Register a new webhook.
...
+                    + by_all_clients
+
+### customFieldDefinitions.add [GET /customFieldDefinitions.add]
+
+Add a custom field definition
+
++ Request (application/json;charset=utf-8)
+
+    + Attributes (object)
+        + id: `20` (number, required)
```
Comment (Brammm) [inline apiary.apib]: Q: is this complete?

File: apiary.apib [hunk 4]
```diff
@@ -2793,6 +2793,180 @@ Register a new webhook.
...
+            + readable (enum[string])
+                + Members
+                    + by_owning_client
+                    + by_all_clients
+            + writable (enum[string])
+                + Members
+                    + by_owning_client
+                    + by_all_clients
+
+### customFieldDefinitions.add [GET /customFieldDefinitions.add]
```
Comment (Brammm) [inline apiary.apib]: PP: should be a POST, no?

File: apiary.apib [hunk 5]
```diff
@@ -2793,6 +2793,180 @@ Register a new webhook.
...
+        + readable (enum[string])
+            + Members
+                + by_owning_client
+                + by_all_clients
+        + writable (enum[string])
+            + Members
+                + by_owning_client
+                + by_all_clients
+
+### customFieldDefinitions.update [GET /customFieldDefinitions.update]
```
Comment (Brammm) [inline apiary.apib]: PP: POST instead of GET

File: apiary.apib (hunk 1 above)
Comment (stivni) [reply apiary.apib]: It should make sense to the (sub)domain or bounded contect...

For timetracking and invoicing, it is the `source`
For custom fields, it is the context
...

File: apiary.apib (hunk 2 above)
Comment (stivni) [reply apiary.apib]: because I still had to look them up, but already wanted to push out the PR, to gather feedback ;)

**Comment by stivni** [comment]
Custom fields have also groups... not sure how to do this yet...
Any ideas?

File: apiary.apib (hunk 3 above)
Comment (stivni) [reply apiary.apib]: I've put the request at the response and vice versa, thanks!

File: apiary.apib [hunk 6]
```diff
@@ -2787,6 +2787,176 @@ Register a new webhook.
...
+                    + auto_increment
+                    + string
+                    + enum
+                    + user
+                    + company
+                    + contact
+                    + ___more___
+            + label (string)
+            + options (array[string])
+            + client - The client who made the integration
```
Comment (stivni) [inline apiary.apib]: `integration` => `custom field definition`

File: apiary.apib (hunk 2 above)
Comment (stivni) [reply apiary.apib]: done

 :ok: by @stivni on 19/10/2017 17:53:28

**Comment by stivni** [comment]
Apparently there is something like "ordering" of custom fields too... (according to DB)
But we can't seem to find where that is manageable...

Also: we need to decide how to cope with `searchable` `options` `can_add_extra_options` and `auto_increment_start` too...

**Comment by stivni** [comment]
will address `change type customField to customFieldDefinition everywhere` in follow up PR

File: apiary.apib [hunk 7]
```diff
@@ -3205,6 +3205,263 @@ Register a new webhook.
...
+                        + product
+                        + invoice
+                        + subscription
+                        + activity - Will be available on both calendar events & todos, as they are convertable to eachother
+                        + ticket
+                + type (enum[string])
+                    + Members
+                        + single_line - Single line text
+                        + multi_line - Multiline text
+                        + single_select - Single selection - has configuration
```
Comment (jenssegers) [inline apiary.apib]: enum?

File: apiary.apib [hunk 8]
```diff
@@ -3205,6 +3205,263 @@ Register a new webhook.
...
+                        + multi_select - Multiple selection - has configuration
```
Comment (jenssegers) [inline apiary.apib]: set?

File: apiary.apib [hunk 9]
```diff
@@ -3205,6 +3205,263 @@ Register a new webhook.
...
+                        + auto_complete - has configuration
```
Comment (jenssegers) [inline apiary.apib]: Auto complete sounds a bit vague to me. Is this a set or enum?

File: apiary.apib [hunk 10]
```diff
@@ -3205,6 +3205,263 @@ Register a new webhook.
...
+                        + contact - Contact
+                        + product - Product
+                        + user - User
+                + label (string)
+                + group (string, nullable)
+                + required: `false` (boolean)
+                + configuration (object, nullable)
+                    + One Of
+                        + options: foo, bar (array[string])
+                        + start_value: 1 (number)
```
Comment (jenssegers) [inline apiary.apib]: initial_value could also be an option

**Pull request merged**